### PR TITLE
feat: Check that all rules have the correct CRS tag and version

### PR DESCRIPTION
**Pull request by airween**
Add two new features for `rules-check.py` linter:
* check that all rules have a `tag` with `OWASP_CRS`
* check that all rules have `ver` action with correct version string

Version can be passed manually with `-v`/`--version`. If no argument given, the script tries to get that from output of `git describe --tags`.

How the version is determined:
* eg `git describe --tags` gives string `v4.5.0-6-g872a90ab`
* `-6-` means there are 6 commits after `4.6.0`, so this is a `dev` version
  * in this case we increment MINOR version: `5` -> `6`
  * and version gets `-dev`
* the result will be: `4.6.0-dev` (`v` will be removed from the start position)
* if the version string is something like this: `v4.5.0-0-abcd01234` then the version will be `4.5.0`

In all cases the `OWASP_CRS/` prefix will be appended.

### CODE REVIEW

**Comment by airween** [comment]
This PR closes #3573.

**Comment by airween** [comment]
@theseion could you help me what's the issue [here](https://github.com/coreruleset/coreruleset/actions/runs/10273153433/job/28426990273?pr=3788#step:8:38). I think there is no `git` command in workflow environment, or I don't know why splitted string contains only 1 item.

**Comment by theseion** [comment]
The checkout action may not fetch any tags. To debug, add the outputs of `git tag -l` and `git fetch --tags` to your script or to the pipeline.

**Comment by airween** [comment]
> The checkout action may not fetch any tags. To debug, add the outputs of `git tag -l` and `git fetch --tags` to your script or to the pipeline.

Hmmm... Could you take a look at [this](https://github.com/coreruleset/coreruleset/actions/runs/10282424385/job/28454097023?pr=3788#step:9:14)?

```
fatal: No tags can describe '3bef193c7d191f8ff9aac25d5548683dd927209e'.
Try --always, or create some tags.
Error: Process completed with exit code 128.
```

**Comment by theseion** [comment]
Fixed.

File: util/crs-rules-check/rules-check.py [hunk 1]
```diff
@@ -343,13 +344,13 @@ def check_tx_variable(self, fname):
...
-                        val_act = re.findall("%\{(tx.[^%]*)\}", a['act_arg'], re.I)
+                        val_act = re.findall(r"%\{(tx.[^%]*)\}", a['act_arg'], re.I)
                     if "act_arg_val" in a and a['act_arg_val'] is not None:
-                        val_act_arg = re.findall("%\{(tx.[^%]*)\}", a['act_arg_val'], re.I)
+                        val_act_arg = re.findall(r"%\{(tx.[^%]*)\}", a['act_arg_val'], re.I)
                     for v in val_act + val_act_arg:
                         v = v.lower().replace("tx.", "")
                         # check whether the variable is a captured var, eg TX.1 - we do not care that case
-                        if not re.match("^\d$", v, re.I):
+                        if not re.match("^\\d$", v, re.I):
```
Comment (theseion) [inline util/crs-rules-check/rules-check.py:353]:
```suggestion
                        if not re.match(r"^\d$", v, re.I):
```

**Comment by airween** [comment]
> Fixed.

Thanks! Seems like all test has passed.

File: util/crs-rules-check/rules-check.py [hunk 2]
```diff
@@ -368,14 +369,14 @@ def check_tx_variable(self, fname):
                     aidx += 1
 
                 if "operator_argument" in d:
-                    oparg = re.findall("%\{(tx.[^%]*)\}", d['operator_argument'], re.I)
+                    oparg = re.findall("%\\{(tx.[^%]*)\\}", d['operator_argument'], re.I)
```
Comment (theseion) [inline util/crs-rules-check/rules-check.py:372]:
```suggestion
                    oparg = re.findall(r"%\{(tx.[^%]*)\}", d['operator_argument'], re.I)
```

File: util/crs-rules-check/rules-check.py [hunk 3]
```diff
@@ -368,14 +369,14 @@ def check_tx_variable(self, fname):
...
                     if oparg:
                         for o in oparg:
                             o = o.lower()
-                            o = re.sub("tx\.", "", o, re.I)
+                            o = re.sub(r"tx\.", "", o, re.I)
                             if (o not in self.globtxvars or phase < self.globtxvars[o]['phase']) and \
-                              not re.match("^\d$", o) and \
-                              not re.match("\/.*\/", o) and \
+                              not re.match(r"^\d$", o) and \
+                              not re.match(r"\/.*\/", o) and \
```
Comment (theseion) [inline util/crs-rules-check/rules-check.py:379]:
```suggestion
                              not re.match(r"/.*/", o) and \
```

File: util/crs-rules-check/rules-check.py [hunk 4]
```diff
@@ -385,8 +386,8 @@ def check_tx_variable(self, fname):
                                     'message': "TX variable '%s' not set / later set (OPARG) in rule %d" % (o, ruleid)
                                 })
                             elif o in self.globtxvars and phase >= self.globtxvars[o]['phase'] and \
-                                not re.match("^\d$", o) and \
-                                not re.match("\/.*\/", o):
+                                not re.match(r"^\d$", o) and \
+                                not re.match(r"\/.*\/", o):
```
Comment (theseion) [inline util/crs-rules-check/rules-check.py:390]:
```suggestion
                                not re.match(r"/.*/", o):
```

File: util/crs-rules-check/rules-check.py [hunk 5]
```diff
@@ -402,8 +403,8 @@ def check_tx_variable(self, fname):
                                 # * rule's phase lower than declaration's phase
                                 rvar = v['variable_part'].lower()
                                 if (rvar not in self.globtxvars or (ruleid != self.globtxvars[rvar]['ruleid'] and phase < self.globtxvars[rvar]['phase'])) and \
-                                  not re.match("^\d$", rvar) and \
-                                  not re.match("\/.*\/", rvar):
+                                  not re.match(r"^\d$", rvar) and \
+                                  not re.match(r"\/.*\/", rvar):
```
Comment (theseion) [inline util/crs-rules-check/rules-check.py:407]:
```suggestion
                                  not re.match(r"/.*/", rvar):
```

File: util/crs-rules-check/rules-check.py [hunk 6]
```diff
@@ -412,8 +413,8 @@ def check_tx_variable(self, fname):
                                         'message': "TX variable '%s' not set / later set (VAR)" % (v['variable_part'])
                                     })
                                 elif rvar in self.globtxvars and phase >= self.globtxvars[rvar]['phase'] and \
-                                    not re.match("^\d$", rvar) and \
-                                    not re.match("\/.*\/", rvar):
+                                    not re.match(r"^\d$", rvar) and \
+                                    not re.match(r"\/.*\/", rvar):
```
Comment (theseion) [inline util/crs-rules-check/rules-check.py:417]:
```suggestion
                                    not re.match(r"/.*/", rvar):
```

File: util/crs-rules-check/rules-check.py [hunk 7]
```diff
@@ -494,9 +495,9 @@ def check_pl_consistency(self):
                             txv = a['act_arg'][3:].split("=")
                             txv[0] = txv[0].lower()                     # variable name
                             if len(txv) > 1:
-                                txv[1] = txv[1].lower().strip("+\{\}")  # variable value
+                                txv[1] = txv[1].lower().strip("+\\{\\}")  # variable value
```
Comment (theseion) [inline util/crs-rules-check/rules-check.py:498]:
```suggestion
                                txv[1] = txv[1].lower().strip(r"+\{\}")  # variable value
```

File: util/crs-rules-check/rules-check.py [hunk 8]
```diff
@@ -494,9 +495,9 @@ def check_pl_consistency(self):
...
                             else:
-                                txv.append(a['act_arg_val'].strip("+\{\}"))
+                                txv.append(a['act_arg_val'].strip("+\\{\\}"))
```
Comment (theseion) [inline util/crs-rules-check/rules-check.py:500]:
```suggestion
                                txv.append(a['act_arg_val'].strip(r"+\{\}"))
```

File: util/crs-rules-check/rules-check.py [hunk 9]
```diff
@@ -535,8 +536,8 @@ def check_pl_consistency(self):
 
                 for t in _txvars:
                     subst_val = re.search("%{tx.[a-z]+_anomaly_score}", _txvars[t], re.I)
-                    val = re.sub("[\+\%\{\}]", "", _txvars[t]).lower()
-                    scorepl = re.search("anomaly_score_pl\d$", t)   # check if last char is a numeric, eg ...anomaly_score_pl1
+                    val = re.sub(r"[\+\%\{\}]", "", _txvars[t]).lower()
```
Comment (theseion) [inline util/crs-rules-check/rules-check.py:539]:
```suggestion
                    val = re.sub(r"[+%{}]", "", _txvars[t]).lower()
```

File: util/crs-rules-check/rules-check.py [hunk 10]
```diff
@@ -627,6 +628,102 @@ def check_lowercase_ignorecase(self):
...
+                        if chainlevel == 0:
+                            if a['act_arg'] == 'OWASP_CRS':
+                                has_crs = True
+                    aidx += 1
+                if ruleid > 0 and has_crs == False:
+                    self.nocrstags.append({
+                        'ruleid' : ruleid,
+                        'line'   : a['lineno'],
+                        'endLine': a['lineno'],
+                        'message': "rule does not have tag with value 'OWASP_CRS'; rule id: %d" % (ruleid)
```
Comment (theseion) [inline util/crs-rules-check/rules-check.py:668]:
```suggestion
                        'message': f"rule does not have tag with value 'OWASP_CRS'; rule id: {ruleid}"
```

File: util/crs-rules-check/rules-check.py [hunk 11]
```diff
@@ -627,6 +628,102 @@ def check_lowercase_ignorecase(self):
...
+                            else:
+                                ruleversion = a['act_arg']
+                    aidx += 1
+                if ruleid > 0 and chainlevel == 0:
+                    if has_ver == False:
+                        self.noveract.append({
+                            'ruleid' : ruleid,
+                            'line'   : a['lineno'],
+                            'endLine': a['lineno'],
+                            'message': "rule does not have 'ver' action; rule id: %d" % (ruleid)
```
Comment (theseion) [inline util/crs-rules-check/rules-check.py:716]:
```suggestion
                            'message': f"rule does not have 'ver' action; rule id: {ruleid}"
```

File: util/crs-rules-check/rules-check.py [hunk 12]
```diff
@@ -627,6 +628,102 @@ def check_lowercase_ignorecase(self):
...
+                        })
+                    else:
+                        if ver_is_ok == False:
+                            self.noveract.append({
+                                'ruleid' : ruleid,
+                                'line'   : a['lineno'],
+                                'endLine': a['lineno'],
+                                'message': "rule's 'ver' action has incorrect value; rule id: %d, version: '%s', expected: '%s'" % (ruleid, ruleversion, crsversion)
```
Comment (theseion) [inline util/crs-rules-check/rules-check.py:724]:
```suggestion
                                'message': f"rule's 'ver' action has incorrect value; rule id: {ruleid}, version: '{ruleversion}', expected: '{crsversion}'"
```

File: util/crs-rules-check/rules-check.py [hunk 13]
```diff
@@ -711,13 +808,31 @@ def msg(msg):
...
 
+def generate_version_string():
+    """
+    generate version string from git tag
+    program calls "git describe --tags" and converts it to version
+    eg:
+      v4.5.0-6-g872a90ab -> "4.6.0-dev"
+      v4.5.0-0-abcd01234 -> "4.5.0"
+    """
+    result = subprocess.run(["git", "describe", "--tags", "--match", "v*.*.*"], capture_output=True, text=True)
```
Comment (theseion) [inline util/crs-rules-check/rules-check.py:819]: I prefer `Popen`. But this works, so it's fine.

**Review by theseion — CHANGES_REQUESTED** [review]

File: util/crs-rules-check/rules-check.py (hunk 13 above)
Comment (airween) [reply util/crs-rules-check/rules-check.py:819]: Usually I use `subprocess` - I should keep this as is now.

**Review by theseion — APPROVED** [review]

**Pull request merged**